packages/flutter_tools/lib/src/protocol_discovery.dart

```diff
@@ -22,9 +24,9 @@ class ProtocolDiscovery {
     assert(logReader != null);
     assert(portForwarder == null || defaultHostPort != null);
     _deviceLogSubscription = logReader.logLines.listen(_handleLine);
-    _timer = new Timer(const Duration(seconds: 60), () {
+    _timer = new Timer(kDiscoveryTimeout, () {
       _stopScrapingLogs();
-      _completer.completeError(new ToolExit('Timeout while attempting to retrieve URL for $serviceName'));
+      _completer.completeError(new ToolExit('Timeout attempting to discover $serviceName URL'));
```

Comment:
> What causes this error to occur?

From what I've seen, extremely slow build steps. 

> What can the user do about it?

Re-try the launch, mostly. Increasing the timeout will mean that users see it less often. I've only seen this trigger within the last month or two, on a very slow (0.9GHz) macbook. We could add verbiage about re-trying the launch in the error message.